src/vsr/replica.zig

```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+                if (received) |message| {
+                    const view_normal = @intCast(u32, message.header.timestamp);
+                    // The view in which this replica's status was normal must be before this view.
+                    assert(view_normal < message.header.view);
+
+                    if (view_normal < view_normal_canonical) continue;
+                    assert(view_normal == view_normal_canonical);
+
+                    const message_headers = message_body_as_headers(message);
+                    for (message_headers) |*header| {
+                        log.debug(
+                            "{}: on_do_view_change: canonical: replica={} op={} checksum={}",
+                            .{
+                                self.replica,
+                                message.header.replica,
+                                header.op,
+                                header.checksum,
+                            },
+                        );
+                    }
+                    self.replace_headers(message_headers);
+                }
             }
 
-            assert(self.op == latest.op);
-            self.journal.remove_entries_from(self.op + 1);
-            assert(self.journal.header_with_op(self.op).?.checksum == latest.checksum);
+            // Since we used do_view_change_head to set the replica.op, it must have been loaded
+            // into the headers (if it wasn't present already).
+            assert(self.journal.header_with_op(self.op) != null);
+
+            // The iteration order of DVCs for repair does not impact the final result.
+            // In other words, you can't end up in a situation with a DVC quorums like:
+            //
+            //   replica     headers  commit_min
+            //         0   4 5 _ _ 8           4 (new leader; handling DVC quorum)
+            //         1   4 _ 6 _ 8           4
+            //         2   4 _ _ 7 8           4
+            //         3  (4 5 6 7 8)          8 (didn't participate in view change)
+            //         4  (4 5 6 7 8)          8 (didn't participate in view change)
+            //
+            // where the new leader's headers depends on which of replica 1 and 2's DVCs is used
+            // for repair before the other (i.e. whether they repair op 6 or 7 first).
+            //
+            // For the above case to occur, replicas 0, 1, and 2 must all share the highest `view_normal`.
+            // And since they share the latest `view_normal`, ops 5,6,7 were just installed by
+            // `replace_header`, which is order-independent (it doesn't use the hash chain).
+            //
+            // (If replica 0's view_normal was greater than 1/2's, then replica 0 must have all
+            // headers from previous views. Which means 6,7 are from the current view. But since
+            // replica 0 doesn't have 6/7, then replica 1/2 must share the latest view_normal. ∎)
```

Comment:
```suggestion
            // replica 0 doesn't have 6/7, then replica 1/2 must share the latest view_normal.)
```

A microdot? 😉

Reply:
Thanks for the awesome worked example!

Reply:
Let's keep it. And then let it be an Easter egg. i.e. We have so many proofs that we won't always use it.